tools/src/main/scala/scala/scalanative/build/Config.scala

```diff
@@ -0,0 +1,211 @@
+package scala.scalanative
+package build
+
+import java.nio.file.{Path, Paths}
+
+import nir.Global
+
+sealed trait Config {
+
+  /** Path to the nativelib jar. */
+  def nativeLib: Path
+
+  /** The driver to use for the optimizer. */
+  def driver: OptimizerDriver
+
+  /** The reporter that receives messages from the linker. */
+  def linkerReporter: LinkerReporter
+
+  /** The reporter that receives messages from the optimizer. */
+  def optimizerReporter: OptimizerReporter
+
+  /** Entry point for linking. */
+  def entry: String
+
+  /** Sequence of all NIR locations. */
+  def paths: Seq[Path]
+
+  /** Directory to emit intermediate compilation results. */
+  def workdir: Path
+
+  /** The path to the `clang` executable. */
+  def clang: Path
+
+  /** The path to the `clang++` executable. */
+  def clangpp: Path
+
+  /** Target triple. */
+  def target: String
+
+  /** The options passed to LLVM's linker. */
+  def linkingOptions: Seq[String]
+
+  /** The compilation options passed to LLVM. */
+  def compileOptions: Seq[String]
+
+  /** The garbage collector to use. */
+  def gc: GarbageCollector
+
+  /** Should stubs be linked? */
+  def linkStubs: Boolean
+
+  /** The logger used by the toolchain. */
+  def logger: Logger
+
+  /** Create a new config with given path to nativelib. */
+  def withNativeLib(value: Path): Config
+
+  /** Create a new config with given driver. */
+  def withDriver(value: OptimizerDriver): Config
+
+  /** Create a new config with given linker reporter. */
+  def withLinkerReporter(value: LinkerReporter): Config
+
+  /** Create a new config with given optimizer reporter. */
+  def withOptimizerReporter(value: OptimizerReporter): Config
+
+  /** Create new config with given entry point. */
+  def withEntry(value: String): Config
+
+  /** Create a new config with given nir paths. */
+  def withPaths(value: Seq[Path]): Config
+
+  /** Create a new config with given directory. */
+  def withWorkdir(value: Path): Config
+
+  /** Create a new config with given path to clang. */
+  def withClang(value: Path): Config
+
+  /** Create a new config with given path to clang++. */
+  def withClangPP(value: Path): Config
+
+  /** Create a new config with given target triple. */
+  def withTarget(value: String): Config
+
+  /** Create a new config with given linking options. */
+  def withLinkingOptions(value: Seq[String]): Config
+
+  /** Create a new config with given compilation options. */
+  def withCompileOptions(value: Seq[String]): Config
+
+  /** Create a new config with given garbage collector. */
+  def withGC(value: GarbageCollector): Config
+
+  /** Create a new config with given behavior for stubs. */
+  def withLinkStubs(value: Boolean): Config
+
+  /** Create a new config with the given logger. */
+  def withLogger(value: Logger): Config
+}
+
+object Config {
+
+  def default(nativeLib: Path,
```

Comment:
Needs a docstring. Additionally, it's not very clear when one should use `empty` and when `default`.